src/compiler/checker.ts

```diff
@@ -25762,10 +25762,11 @@ namespace ts {
                         if (!(links.flags & NodeCheckFlags.InCheckIdentifier)) {
                             links.flags |= NodeCheckFlags.InCheckIdentifier;
                             const parentType = getTypeForBindingElementParent(parent, CheckMode.Normal);

```

Comment:
We are receiving `T` here from here:
https://github.dev/microsoft/TypeScript/blob/71e852922888337ef51a0e48416034a94a6c34d9/src/compiler/checker.ts#L9030-L9034

I've compared the code run for the second case - where those variabeld were destructured into a const declaration in the function's body and not directly in the parameter position and I've discovered that in that case the returned type was `Action`, from here:
https://github.dev/microsoft/TypeScript/blob/71e852922888337ef51a0e48416034a94a6c34d9/src/compiler/checker.ts#L9088-L9089
That in turn could be traced back to `getNarrowableTypeForReference` here:
https://github.dev/microsoft/TypeScript/blob/71e852922888337ef51a0e48416034a94a6c34d9/src/compiler/checker.ts#L26001

So I've just replicated the logic from there (just slightly simplified):
https://github.dev/microsoft/TypeScript/blob/71e852922888337ef51a0e48416034a94a6c34d9/src/compiler/checker.ts#L25747

In here, we are always interested in narrowable types so it made sense to me to just unconditionally resolve through `getBaseConstraintOrType`